### PR TITLE
EIP/RIP-7212

### PR DESCRIPTION
**Pull request by alexb5dh**
## Changes

Adds secp256r1 precompile contract. 
For now intended to be used in Optimism, but may be enabled in Ethereum as agreed.

[EIP](https://eips.ethereum.org/EIPS/eip-7212) | [RIP](https://github.com/ethereum/RIPs/blob/master/RIPS/rip-7212.md) | [Discussion](https://ethereum-magicians.org/t/eip-7212-precompiled-for-secp256r1-curve-support/14789) | [Geth PR](https://github.com/ethereum/go-ethereum/pull/27540)

#### What types of changes does your code introduce?

- [x] New feature (a non-breaking change that adds functionality)

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by alexb5dh** [comment]
Current implementation uses built-in .NET ECDsa, which is not the fastest option. Benchmarks comparing different versions:

## .NET

### Windows 10

| Method                             | Mean      | Error    | StdDev   |
|----------------------------------- |----------:|---------:|---------:|
| DotNet_Ecdsa_WithCreate            | 436.08 us | 4.865 us | 6.819 us |
| DotNet_Ecdsa_WithImportParameters  | 448.28 us | 4.319 us | 6.330 us |
| BouncyCastle                       | 227.36 us | 3.874 us | 5.798 us |
| Go_Ecdsa_MarshallingString         |  68.06 us | 0.399 us | 0.572 us |
| Go_Ecdsa_MarshallingSlice          |  66.98 us | 0.346 us | 0.507 us |

### Ubuntu 22

| Method                            | Mean      | Error    | StdDev   |
|---------------------------------- |----------:|---------:|---------:|
| DotNet_Ecdsa_WithCreate           | 144.05 us | 1.038 us | 1.488 us |
| DotNet_Ecdsa_WithImportParameters | 142.64 us | 0.755 us | 1.107 us |
| BouncyCastle                      | 208.14 us | 1.339 us | 1.962 us |
| Go_Ecdsa_MarshallingString        |  68.89 us | 0.356 us | 0.533 us |
| Go_Ecdsa_MarshallingSlice         |  69.93 us | 1.772 us | 2.365 us |

## Go 

### Windows 10

| Method | Mean     |
|--------|---------:|
| Verify | 66.05 us |

### Ubuntu 22

| Method | Mean     |
|--------|---------:|
| Verify | 64.92 us |

**Comment by benaadams** [comment]
Which go lib specifically is that?

**Comment by alexb5dh** [comment]
> Which go lib specifically is that?

@benaadams, built-in [ecdsa package](https://pkg.go.dev/crypto/ecdsa#Verify).

**Comment by alexb5dh** [comment]
C# benchmarking code:
```csharp
public class Secp256r1VerifyBenchmarks
{
    private const string InputHex = "4cee90eb86eaa050036147a12d49004b6b9c72bd725d39d4785011fe190f0b4da73bd4903f0ce3b639bbbf6e8e80d16931ff4bcf5993d58468e8fb19086e8cac36dbcd03009df8c59286b162af3bd7fcc0450c9aa81be5d10d312af6c66b1d604aebd3099c618202fcfe16ae7770b0c49ab5eadf74b754204a3bb6060e44eff37618b065f9832de4ca6ca971a7a1adc826d0f7c00181a5fb2ddf79ae00b4e10e";
    private static readonly byte[] Input = Convert.FromHexString(InputHex);

    private ECDsa _ecdsaShared;
    private X9ECParameters _bouncyCurve;
    private ISigner _bouncySigner;
    private ECDomainParameters _bouncyDomainParameters;

    [GlobalSetup]
    public void Setup()
    {
        _ecdsaShared = ECDsa.Create();

        _bouncyCurve = CustomNamedCurves.GetByName("secp256r1");
        _bouncySigner = SignerUtilities.GetSigner("NONEWITHPLAIN-ECDSA");
        _bouncyDomainParameters = new(_bouncyCurve);
    }

    [Benchmark]
    public bool DotNet_Ecdsa_WithCreate()
    {
        ReadOnlySpan<byte> input = Input;
        ReadOnlySpan<byte> hash = input[..32];
        ReadOnlySpan<byte> sig = input[32..96];
        ReadOnlySpan<byte> x = input[96..128];
        ReadOnlySpan<byte> y = input[128..160];

        var ecdsa = ECDsa.Create(new ECParameters
        {
            Curve = ECCurve.NamedCurves.nistP256,
            Q = new() { X = x.ToArray(), Y = y.ToArray() }
        });

        return ecdsa.VerifyHash(hash, sig);
    }

    [Benchmark]
    public bool DotNet_Ecdsa_WithImportParameters()
    {
        ReadOnlySpan<byte> input = Input;
        ReadOnlySpan<byte> hash = input[..32];
        ReadOnlySpan<byte> sig = input[32..96];
        ReadOnlySpan<byte> x = input[96..128];
        ReadOnlySpan<byte> y = input[128..160];

        ECDsa ecdsa = _ecdsaShared;
        ecdsa.ImportParameters(new ECParameters
        {
            Curve = ECCurve.NamedCurves.nistP256,
            Q = new() { X = x.ToArray(), Y = y.ToArray() }
        });

        return ecdsa.VerifyHash(hash, sig);
    }

    [Benchmark] // 211.2 us
    public bool BouncyCastle()
    {
        ReadOnlySpan<byte> input = Input;
        ReadOnlySpan<byte> hash = input[..32];
        ReadOnlySpan<byte> sig = input[32..96];
        ReadOnlySpan<byte> x = input[96..128];
        ReadOnlySpan<byte> y = input[128..160];

        ECPublicKeyParameters parameters = new(
            "ECDSA",
            _bouncyCurve.Curve.CreatePoint(new Org.BouncyCastle.Math.BigInteger(1, x), new Org.BouncyCastle.Math.BigInteger(1, y)),
            _bouncyDomainParameters
        );

        _bouncySigner.Init(false, parameters);
        _bouncySigner.BlockUpdate(hash);
        return _bouncySigner.VerifySignature(sig.ToArray());
    }

#if Windows
    [DllImport("secp256r1.dll", CallingConvention = CallingConvention.Cdecl)]
#elif Linux
    [DllImport("secp256r1.so", CallingConvention = CallingConvention.Cdecl)]
#endif
    private static extern byte VerifyHex(string hash);

    [Benchmark]
    public bool Go_Ecdsa_MarshallingString() => VerifyHex(InputHex) != 0;

#if Windows
    [DllImport("secp256r1.dll", CallingConvention = CallingConvention.Cdecl)]
#elif Linux
    [DllImport("secp256r1.so", CallingConvention = CallingConvention.Cdecl)]
#endif
    private static extern byte VerifyBytes(GoSlice hash);

    private struct GoSlice(IntPtr data, long len, long cap)
    {
        public IntPtr Data = data;
        public long Len = len, Cap = cap;
    }

    [Benchmark]
    public bool Go_Ecdsa_MarshallingSlice()
    {
        unsafe
        {
            fixed (byte* p = Input)
            {
                var ptr = (IntPtr) p;
                var slice = new GoSlice(ptr, Input.Length, Input.Length);
                return VerifyBytes(slice) != 0;
            }
        }
    }

    // Other options:
    // https://github.com/starkbank/ecdsa-dotnet - needs bridge/pinvoke from Rust
    // https://github.com/MystenLabs/fastcrypto - very slow actually
}
```

**Review by smartprogrammer93 — APPROVED** [review]

**Review by flcl42 — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/Precompiles/Secp256r1Precompile.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    public long BaseGasCost(IReleaseSpec releaseSpec) => 3450L;
+    public long DataGasCost(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec) => 0L;
+
+    // TODO can be optimized - Go implementation is 2-6 times faster depending on the platform. Options:
+    // - Try to replicate Go version in C#
+    // - Compile Go code into a library and call it via P/Invoke
+    public (ReadOnlyMemory<byte>, bool) Run(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec)
+    {
+        if (inputData.Length != 160)
+            return (null, true);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Precompiles/Secp256r1Precompile.cs:28]: should it be `(null, true)` or `IPrecompile.Failure`?

File: src/Nethermind/Nethermind.Evm/Precompiles/Secp256r1Precompile.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        using var ecdsa = ECDsa.Create(new ECParameters
+        {
+            Curve = ECCurve.NamedCurves.nistP256,
+            Q = new() { X = x.ToArray(), Y = y.ToArray() }
+        });
+        var isValid = ecdsa.VerifyHash(hash, sig);
+
+        Metrics.Secp256r1Precompile++;
+
+        return (isValid ? ValidResult : null, true);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Precompiles/Secp256r1Precompile.cs:43]: same here for invalid

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/Precompiles/Secp256r1Precompile.cs (hunk 1 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Evm/Precompiles/Secp256r1Precompile.cs:28]: They decided to return always true (and did not spec it well) and now do not want to change: https://ethereum-magicians.org/t/eip-7212-precompiled-for-secp256r1-curve-support/14789/104

**Comment by benaadams** [comment]
Aside; this is different to the `SecP256k1` lib we use for Ecdsa sign and recover?

**Comment by alexb5dh** [comment]
It's for secP256_r_1

**Comment by alexb5dh** [comment]
Noticed that Go implementation does some caching, at least [when converting public key to C version](https://github.com/golang/go/blob/master/src/crypto/ecdsa/boring.go#L16), so updated benchmarks to try to eliminate/reduce any caching benefits. Didn't find any noticeable difference so:

```csharp
public class Secp256r1VerifyBenchmarks
{
    public byte[][] Inputs;
    public string[] InputsHexes;

    private ECDsa _ecdsaShared;
    private X9ECParameters _bouncyCurve;
    private ISigner _bouncySigner;
    private ECDomainParameters _bouncyDomainParameters;

    [Params(1, 100, 10_000)]
    public int SamplesCount;

    [GlobalSetup]
    public void Setup()
    {
        _ecdsaShared = ECDsa.Create();

        _bouncyCurve = CustomNamedCurves.GetByName("secp256r1");
        _bouncySigner = SignerUtilities.GetSigner("NONEWITHPLAIN-ECDSA");
        _bouncyDomainParameters = new(_bouncyCurve);

        var rng = RandomNumberGenerator.Create();
        Inputs = new byte[SamplesCount][];
        InputsHexes = new string[SamplesCount];
        for (var i = 0; i < Inputs.Length; i++)
        {
            var hash = new byte[32];
            rng.GetBytes(hash);

            var ecdsa = ECDsa.Create(ECCurve.NamedCurves.nistP256);
            ECParameters pub = ecdsa.ExportParameters(false);
            byte[] sig = ecdsa.SignHash(hash, DSASignatureFormat.IeeeP1363FixedFieldConcatenation);
            (byte[] x, byte[] y) = (pub.Q.X, pub.Q.Y);
            Inputs[i] = [..hash, ..sig, ..x, ..y];
            InputsHexes[i] = Convert.ToHexString(Inputs[i]);
        }
    }

    private int Index => Random.Shared.Next(Inputs.Length);
    private byte[] Input => Inputs[Index];
    private string InputHex => InputsHexes[Index];

    [Benchmark]
    public bool DotNet_Ecdsa_WithCreate()
    {
        ReadOnlySpan<byte> input = Input;
        ReadOnlySpan<byte> hash = input[..32];
        ReadOnlySpan<byte> sig = input[32..96];
        ReadOnlySpan<byte> x = input[96..128];
        ReadOnlySpan<byte> y = input[128..160];

        var ecdsa = ECDsa.Create(new ECParameters
        {
            Curve = ECCurve.NamedCurves.nistP256,
            Q = new() { X = x.ToArray(), Y = y.ToArray() }
        });

        return ecdsa.VerifyHash(hash, sig);
    }

    [Benchmark]
    public bool DotNet_Ecdsa_WithImportParameters()
    {
        ReadOnlySpan<byte> input = Input;
        ReadOnlySpan<byte> hash = input[..32];
        ReadOnlySpan<byte> sig = input[32..96];
        ReadOnlySpan<byte> x = input[96..128];
        ReadOnlySpan<byte> y = input[128..160];

        ECDsa ecdsa = _ecdsaShared;
        ecdsa.ImportParameters(new ECParameters
        {
            Curve = ECCurve.NamedCurves.nistP256,
            Q = new() { X = x.ToArray(), Y = y.ToArray() }
        });

        return ecdsa.VerifyHash(hash, sig);
    }

    [Benchmark]
    public bool BouncyCastle()
    {
        ReadOnlySpan<byte> input = Input;
        ReadOnlySpan<byte> hash = input[..32];
        ReadOnlySpan<byte> sig = input[32..96];
        ReadOnlySpan<byte> x = input[96..128];
        ReadOnlySpan<byte> y = input[128..160];

        ECPublicKeyParameters parameters = new(
            "ECDSA",
            _bouncyCurve.Curve.CreatePoint(new Org.BouncyCastle.Math.BigInteger(1, x), new Org.BouncyCastle.Math.BigInteger(1, y)),
            _bouncyDomainParameters
        );

        _bouncySigner.Init(false, parameters);
        _bouncySigner.BlockUpdate(hash);
        return _bouncySigner.VerifySignature(sig.ToArray());
    }

#if Windows
    [DllImport("secp256r1.dll", CallingConvention = CallingConvention.Cdecl)]
#elif Linux
    [DllImport("secp256r1.so", CallingConvention = CallingConvention.Cdecl)]
#endif
    private static extern byte VerifyHex(string hash);

    [Benchmark]
    public bool Go_Ecdsa_MarshallingString() => VerifyHex(InputHex) != 0;

#if Windows
    [DllImport("secp256r1.dll", CallingConvention = CallingConvention.Cdecl)]
#elif Linux
    [DllImport("secp256r1.so", CallingConvention = CallingConvention.Cdecl)]
#endif
    private static extern byte VerifyBytes(GoSlice hash);

    private struct GoSlice(IntPtr data, long len, long cap)
    {
        public IntPtr Data = data;
        public long Len = len, Cap = cap;
    }

    [Benchmark]
    public bool Go_Ecdsa_MarshallingSlice()
    {
        ReadOnlySpan<byte> input = Input;

        unsafe
        {
            fixed (byte* p = input)
            {
                var ptr = (IntPtr) p;
                var slice = new GoSlice(ptr, input.Length, input.Length);
                return VerifyBytes(slice) != 0;
            }
        }
    }

    // Other options:
    // https://github.com/starkbank/ecdsa-dotnet - needs bridge/pinvoke from Rust
    // https://github.com/MystenLabs/fastcrypto - very slow actually
}
```

**Pull request merged**